src/lib.rs

```diff
@@ -68,118 +63,90 @@ enum ParseState {
 impl Forth {
     pub fn new() -> Forth {
         Forth {
-            word_map: WORD_MAP.clone(),
-            stack: VecDeque::new(),
+            word_map: default_word_map(),
+            stack: Vec::new(),
         }
     }
 
     pub fn format_stack(&self) -> String {
-        let mut stack_str = String::new();
-        for v in self.stack.iter() {
-            stack_str.push_str(&v.to_string());
-            stack_str.push_str(" ");
-        }
-        stack_str = stack_str.trim().to_owned();
-        stack_str
+        StackFormat(self).to_string()
     }
 
     pub fn eval(&mut self, input: &str) -> ForthResult {
-        match self.input_parse(input) {
-            Ok(v) => {
-                for i in v.into_iter() {
-                    match i {
-                        Item::Exec_(s) => match s {
-                            Exec::Arith_(o) => {
-                                let (a, b) = match (self.stack.pop_back(), self.stack.pop_back()) {
-                                    (Some(a), Some(b)) => (a, b),
-                                    (_, _) => return Err(Error::StackUnderflow),
-                                };
-                                match eval_oper(a, b, o) {
-                                    Ok(v) => self.stack.push_back(v),
-                                    Err(e) => return Err(e),
-                                }
-                            },
-                            Exec::Stack_(c) => {
-                                try!(eval_command(&mut self.stack, c));
-                            },
-                            Exec::Value_(v) => {
-                                self.stack.push_back(v);
-                            },
-                        },
-                        _ => (),
-                    }
+        let v = try!(self.input_parse(input));
+        for i in v {
+            if let Item::Exec_(s) = i {
+                match s {
+                    Exec::Arith_(o) => {
+                        let (a, b) = match (self.stack.pop(), self.stack.pop()) {
+                            (Some(a), Some(b)) => (a, b),
+                            (_, _) => return Err(Error::StackUnderflow),
+                        };
+                        let v = try!(eval_oper(a, b, o));
```

Comment:
These two lines should be brought up into the `Some, Some` case, no need to have temp variables.